relay-server/src/actors/spooler/mod.rs

```diff
@@ -636,17 +727,51 @@ impl BufferState {
                         "failed to spool the in-memory buffer to disk",
                     );
                 }
+                relay_log::trace!(
+                    "Transition to disk spool: # of envelopes = {}",
+                    disk.count.unwrap_or_default()
+                );
+
                 Self::Disk(disk)
             }
-            Self::Disk(disk) if disk.is_empty().await.unwrap_or_default() => {
-                Self::MemoryFileStandby {
-                    ram: InMemory::new(config.spool_envelopes_max_memory_size()),
-                    disk,
+            Self::Disk(mut disk) if Self::is_below_low_mem_watermark(config, &disk).await => {
+                match disk.delete_and_fetch_all(services).await {
+                    Ok(buffer) => {
+                        let ram = InMemory::new_with_buffer(
+                            config.spool_envelopes_max_memory_size(),
+                            buffer,
+                        );
+                        relay_log::trace!(
+                            "Transition to memory spool: # of envelopes = {}",
+                            ram.envelope_count
+                        );
+
+                        Self::MemoryFileStandby { ram, disk }
+                    }
+                    Err(err) => {
+                        relay_log::error!(
+                            error = &err as &dyn Error,
+                            "failed to move data from disk to memory, keep using on-disk spool",
+                        );
+
+                        Self::Disk(disk)
+                    }
                 }
             }
             Self::Memory(_) | Self::MemoryFileStandby { .. } | Self::Disk(_) => self,
         }
     }
+
+    /// Returns `true` if the on-disk spooled data can fit in the memory.
+    ///
+    /// The spooled envelopes must:
+    /// * fit into memory and take not more than 30% of the configured space
+    /// * the used buffer guards also must be under the low watermark.
+    async fn is_below_low_mem_watermark(config: &Config, disk: &OnDisk) -> bool {
+        ((config.spool_envelopes_max_memory_size() as f64 * 0.3) as i64)
```

Comment:
maybe ratio could be a const

Reply:
done in [`b2b35f2`](https://github.com/getsentry/relay/pull/2949/commits/b2b35f2b50f796378f0b84da90c7061eaa9474b6)